core/src/maxicode/MCDecoder.cpp

```diff
@@ -122,25 +122,31 @@ static int GetBit(int bit, const ByteArray& bytes)
 	return (bytes[bit / 6] & (1 << (5 - (bit % 6)))) == 0 ? 0 : 1;
 }
 
-static int GetInt(const ByteArray& bytes, const ByteArray& x)
+static unsigned int GetInt(const ByteArray& bytes, const ByteArray& x)
 {
 	int len = Size(x);
-	int val = 0;
+	unsigned int val = 0;
 	for (int i = 0; i < len; i++)
 		val += GetBit(x[i], bytes) << (len - i - 1);
 
 	return val;
 }
 
-static int GetPostCode2(const ByteArray& bytes)
+static unsigned int GetPostCode2Length(const ByteArray& bytes)
 {
-	return GetInt(bytes,
-				  {33, 34, 35, 36, 25, 26, 27, 28, 29, 30, 19, 20, 21, 22, 23, 24, 13, 14, 15, 16, 17, 18, 7, 8, 9, 10, 11, 12, 1, 2});
+	return std::min(GetInt(bytes, {39, 40, 41, 42, 31, 32}), 9U);
 }
 
-static int GetPostCode2Length(const ByteArray& bytes)
+static std::string GetPostCode2(const ByteArray& bytes)
 {
-	return GetInt(bytes, {39, 40, 41, 42, 31, 32});
+	unsigned int val = GetInt(bytes,
+				  {33, 34, 35, 36, 25, 26, 27, 28, 29, 30, 19, 20, 21, 22, 23, 24, 13, 14, 15, 16, 17, 18, 7, 8, 9, 10, 11, 12, 1, 2});
+	unsigned int len = GetPostCode2Length(bytes);
+	// Pad or truncate to length
+	char buf[11]; // 30 bits 0x3FFFFFFF == 1073741823 (10 digits)
```

Comment:
Any reason why you did not use the (relatively new) ` return ToString(val, len);` here?

Reply:
The issue is the number could be 10 digits wide but only want to 0-fill it to `len`, so unless an extra option were added to `ToString()` (which would seem excessive given how unusual the requirement is) it's not convenient to use it here and this old-skool approach seems simplest.

Reply:
You mean (`val, len)` could be e.g. `(999, 2)` and the expected result in that case would be `999`?

Reply:
The expected result as implemented would be `99` as it's choosing to favour the length. This is all just defending against improper data. (It would be good if there was some mechanism via `Result` to indicate such situations, e.g. a `warnings()` API or something, which I think I recall you mentioning somewhere perhaps.)

Reply:
I see. `ToString(999,2)` would throw a `FormatError`, which I believe would be at least Ok if not the right thing in this situation? Do you have a specific 'valid' input that you are trying to make the library accept here?

Reply:
I noticed the issue on merging master to the "diagnostics2" branch after the change to `ToString()` when running `ZXingReader` on a big directory I have full on all sorts of barcodes from random sources and untrapped exceptions occurred on some files. One such is attached, the postcode data is Postcode "339010000", Postcode Length 5. Where I got this barcode or how real or 'valid' it is I don't know I'm afraid.

![maxicode_sbt_fig5](https://user-images.githubusercontent.com/481982/181464775-4b83b4a6-1b4f-47ad-8634-0e2696ca6d73.png)

Reply:
I tested this sample with other detectors and they seem to be fine with it and report the postcode back as "339010000". I imagine the postcode that the author actually meant to encode was "33901", which is 5 digits long. All of the following options on how to interpret this would sound plausible to me a) it is invalid b) the result is meant to be "33901" and c) the result is meant to be "339010000".

I find the spec text to be unclear regarding the postal code length. Annext B1.4.a says "The remaining 30 bits is the binary representation of the numeric postal code." To my understanding that means the code "33901" can not be encoded as "339010000" and the latter needs to be interpreted as a 9 digit postal code. But what is the function of the length then?

Reply:
Yes it's a very odd interpretation to encode as this sample does - I found where I got the original from -https://www.silverbaytech.com/encoders/maxicode-encoder/. The address and postcode "33901" are real and match, so in this case b) works.Thanks for reviewing & merging.

Reply:
> so in this case b) works

But your code would return it as c), right? I just wanted to actually execute and found that the symbol is not detected due to a checksum error?!? Did I break something else in the last couple days?

Reply:
Not that I can see, it reads for me using master with (from build dir) `example/ZXingReader -escape <img.png>`, returning
`"[)><RS>01<GS>9633901<GS>840<GS>001<GS>1Z34567890<GS>UPSN<GS>102562<GS>034<GS><GS>1/1<GS><GS>Y<GS>2201 Second St<GS>Ft Myers<GS>FL<RS><EOT>"`, i.e. "33901" as postcode.

Reply:
Turns out it was a local change that seemed totally unrelated... so your code does return b) which makes most sense but is not what other decoders return. If only we had a standard that would tell what is and what is not to be expected here ;).